### PR TITLE
Fix WASI fd_readdir on big-endian

### PR DESCRIPTION
**Pull request by uweigand**
This code assumes that the Dirent structure has the same memory
layout on the host (Rust code) as in wasm code.  This is not true
if the host is big-endian, as wasm is always little-endian.

Fixed by always byte-swapping Dirent fields to little-endian
before passing them on to wasm code.

### CODE REVIEW

File: crates/wasi-common/src/snapshots/preview_1.rs [hunk 1]
```diff
@@ -1554,7 +1554,13 @@ fn dirent_bytes(dirent: types::Dirent) -> Vec<u8> {
     let mut bytes = Vec::with_capacity(size);
     bytes.resize(size, 0);
     let ptr = bytes.as_mut_ptr() as *mut types::Dirent;
-    unsafe { ptr.write_unaligned(dirent) };
+    let guest_dirent = types::Dirent {
+        d_ino: dirent.d_ino.to_le(),
+        d_namlen: dirent.d_namlen.to_le(),
+        d_type: dirent.d_type, // endian-invariant
```
Comment (alexcrichton) [inline crates/wasi-common/src/snapshots/preview_1.rs:1560]: Could this assert that the size of this field is 1 so the endianness is known to not matter?

**Review by alexcrichton — APPROVED** [review]
This is definitely a preexisting bug in wiggle that it never tried to handle endidanness, and this is something that we'll ideally be fixing at the bindings generation layer since in general host-code shouldn't have to care about endianness. I suspect this isn't the last endian bug in the WASI bindings, but I'm hoping that we can survive long enough to get to the point where we can properly fix this.

FWIW the semi-replacement of wiggle (more like the next stage in evolution) at https://github.com/bytecodealliance/witx-bindgen/ should fix issues like this.

File: crates/wasi-common/src/snapshots/preview_1.rs (hunk 1 above)
Comment (uweigand) [reply crates/wasi-common/src/snapshots/preview_1.rs:1565]: Like this?   Patch updated.

**Comment by alexcrichton** [comment]
👍

**Pull request merged**